[AIEX] Make all tables in `AIEXGenFormats.inc` constexpr

mludevid opened the pull request:
The main goal of this PR is to speed up the build time of the compiler. These are the build time improvements for `AIEXMCFormats.cpp`:
- AIE1: 5.9s => 2.5s
- AIE2: 8.3s => 3s
- AIE2P: 12.8s => 3.9s

martien-de-jong commented on llvm/lib/Target/AIE/MCTargetDesc/AIEMCFormats.cpp:
```diff
@@ -50,7 +50,7 @@ const MCSlotKind AIEInstFormat::getSlot() const {
   // value.
   // NOTE: so... Overkill to use a map in this case?
```
Technically, it's not a map anymore.

mludevid replied:
I removed the comment.

martien-de-jong commented on llvm/lib/Target/AIE/MCTargetDesc/AIEMCFormats.h:
```diff
@@ -168,9 +174,7 @@ class MCFormatDesc {
-  using SlotsFieldsMap =
-      std::unordered_map<const MCSlotKind, const MCFormatField *,
-                         MCSlotKind::Hasher>;
+  using SlotsFieldsMap = llvm::ArrayRef<SlotFieldPair>;
```
Perhaps you could wrap the ArrayRef in a class with an `at(SlotKind Slot)` method. Perhaps llvm has a suitable container in ADT?

mludevid replied:
I wrapped the ArrayRef with a `SlotsFieldsMap` class.

martien-de-jong commented on llvm/utils/TableGen/CodeGenFormat.cpp:
```diff
@@ -682,51 +684,49 @@ void TGInstrLayout::emitFormat(ConstTable &FieldsHierarchy, ConstTable &o,
-  for (unsigned Idx = 0; Idx <= MaxIdx; Idx++) {
-    FieldRanges << OpFields.ref(Idx);
+  unsigned OpFieldIdx = 0;
+  for (unsigned OperandIdx = 0; OperandIdx <= NumOperands; OperandIdx++) {
```
CHECK: we need to split OpFieldIdx and OperandIdx since they are not 1-1. We have a NFC name change from MaxIdx to NumOperands.

mludevid replied:
Exactly. The name changes were for readability. I could move them into an NFC commit if you want.

martien-de-jong replied:
No, fine.

martien-de-jong commented on llvm/utils/TableGen/CodeGenFormat.h:
```diff
@@ -85,6 +87,14 @@ class ConstTable {
   std::string ref(unsigned Idx) const { return absRef(Mark + Idx); }
   // Make a reference to the next entry
   std::string refNext() const { return absRef(Size); }
+  // Return an ArrayRef of the entries sinze the last marked block
```
nit: since

martien-de-jong commented on llvm/utils/TableGen/CodeGenFormat.h:
```diff
@@ -85,6 +87,14 @@ class ConstTable {
   std::string refNext() const { return absRef(Size); }
+  // Return an ArrayRef of the entries sinze the last marked block
+  std::string arrayRef() const {
+    if (Size - Mark == 0) {
```
Declare a name for `Size - Mark`?

martien-de-jong commented:
Perhaps add in the commit message that we need to make the tables constexpr to avoid a huge static constructor routine that blows up compiler time.

martien-de-jong reviewed (dismissed):
Mostly nits. I could also address them in my followup.

mludevid commented:
I addressed all the comments and updated the commit message to explain the motivation behind this PR.

mludevid committed:
[AIEX] Make all tables in `AIEXGenFormats.inc` constexpr

We needed to make the tables constexpr to avoid a huge static
constructor routine that was blowing up the compilation time of the
compiler.
